pkg/etcdcli/etcdcli.go

```diff
@@ -329,9 +317,63 @@ func (g *etcdClientGetter) UnhealthyMembers() ([]*etcdserverpb.Member, error) {
 		g.eventRecorder.Warningf("UnhealthyEtcdMember", "unhealthy members: %v", strings.Join(unhealthyMemberNames, ","))
 	}
 
+	_, unhealthyMembers := etcdMemberHealth.MemberStatus()
+
 	return unhealthyMembers, nil
 }
 
+func (g *etcdClientGetter) MemberHealth(etcdMembers []*etcdserverpb.Member) (*memberHealth, error) {
+	memberHealth, err := g.getMemberHealth(etcdMembers)
+	if err != nil {
+		return nil, err
+	}
+	return memberHealth, nil
+}
+
+func (g *etcdClientGetter) getMemberHealth(etcdMembers []*etcdserverpb.Member) (*memberHealth, error) {
+	var wg sync.WaitGroup
+	memberHealth := &memberHealth{}
+	hch := make(chan healthCheck, len(etcdMembers))
+	for _, member := range etcdMembers {
+		if len(member.ClientURLs) == 0 {
+			memberHealth.Check = append(memberHealth.Check, healthCheck{Member: member, Health: false, Started: false})
+			continue
+		}
+		wg.Add(1)
+		go func(member *etcdserverpb.Member) {
+			defer wg.Done()
+			cli, err := getEtcdClient([]string{member.ClientURLs[0]})
+			if err != nil {
+				hch <- healthCheck{Member: member, Health: false, Started: true, Error: err.Error()}
+				return
+			}
+			defer cli.Close()
+			st := time.Now()
+			ctx, cancel := context.WithCancel(context.Background())
+			_, err = cli.Get(ctx, "health")
+			cancel()
```

Comment:
(Aside: can somebody explain this context pattern to me? I've seen it throughout and wonder how, as used, it's different than `context.TODO()` as it doesn't seem like we're actually handling cancellation asynchronously or through passthrough of an outer context)